draft/ZEP0003.md

```diff
@@ -0,0 +1,115 @@
+---
+layout: default
+title: ZEP0000
+description: Variable chunk sizes
+parent: template
+nav_order: 1
+---
+
+# ZEP 3 — Variable chunking
+
+---
+
+Authors:
+* Martin Durant ([@martindurant](https://github.com/martindurant)), Anaconda, Inc.
+
+Status: Draft
+
+Type: Specification
+
+Created: 2022-10-17
+
+## Abstract
+
+To allow the chunks of a zarr array to be hyperrectangles rather than hypercubes, with the chunk
+lengths along any dimension a list of integers rather than a single chunk size.
+
+## Motivation and Scope
+
+Two specific use cases have motivated this, given below. However, this generalisation of Zarr's storage
+model can be seen as an optional enhancement, and the same data model as currently used by dask.array.
+
+- when producing a [kerchunked](https://github.com/fsspec/kerchunk) dataset, the native chunking of the targets 
+cannot be changed. It is common
+to have non-regular chunking on one dimension, such as a time dimension with one sample per day and chunks 
+of one month or one year. The change would allow these datasets to be read via kerchunk, and/or converted to
+zarr with equivalent chunking to the original
+- [awkward](https://github.com/scikit-hep/awkward) arrays, ragged arrays and sparse data can be represented as
+a set of one-dimensional arrays, with an appropriate metadata description convention. The size of a chunks
+of each component array corresponding to a logical chunk of the overall array will not, in general be equal
+with each other in a single chunk, nor consistent between chunks.
+- in some cases, parts of the overall data array may have very different data distributions, and it can
+be very convenient to partition the data by such characteristics to allow, for example, for more efficient encoding
+schemes.
+
+## Usage and Impact
+
+
+
+## Backward Compatibility
+
+
+This change is fully backward compatible - all old data will remain usable. However, data written with
+variable chunks will not be readable by older versions of Zarr.
+
+## Detailed description
+
+Currently, array metadata looks something like
+```json
```

Comment:
This is correct for zarr v2 but does not match the current state of zarr v3.  I think it would be helpful to revise this to apply to zarr v3, since I assume your intent here is to propose this as a zarr v3 extension.

The current zarr v3 spec seems to be designed to make the chunking scheme itself an extension point:

https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#chunk-grid

However, that extension point design has not really been discussed at all, nor have any alternative grid types been proposed, other than this one.

This proposal could simply be a new chunk_grid "type".  But I am not sure if that is the best fit --- this proposal allows the non-uniform chunking to be specified for just some of the dimensions.  Additionally, the v3 spec has "separator" as a field of the chunk grid type, but it equally applies to both regular and rectilinear grids.

It might be worth considering if there are any other grid types that would be useful, and if so, how they might interact with this proposal.

Reply:
Yes, totally agree that, as far as the zarr-specs goes, a new subsection in that area is the way to go - with the regular grid being a special case. However, the spec does not go as far as to explicitly define how the chunks are specified in the metadata, it is only descriptive text. Is there a jsonschema somewhere I should propose a change to? When coming to write this ZEP, that's what I had assumed would be the case, but I cannot find one.

The separator is not material in this proposal - any separator can be used equally with regular or irregular grids.

> if there are any other grid types that would be useful

Certainly worth thinking about. Sharding, for instance, would be thought of as hierarchical chunking. I don't know how we can fill it into this proposal, though. Ideas very welcome!

Reply:
> Certainly worth thinking about. Sharding, for instance, would be thought of as hierarchical chunking. I don't know how we can fill it into this proposal, though. Ideas very welcome!

I think it could certainly make sense to use sharding in conjunction with a rectilinear grid --- in fact we could allow both the chunk grid and the shard grid to be rectilinear rather than regular.

Reply:
@jstriebel Any thoughts on this?

Reply:
I have referenced that link and stated that that's where the change will happen.

Reply:
>  if sharding should allow to combinae a flexible number of chunks, but that might be added later if the need arises

This need already arose and has been implemented in the special handling in preffs, an alternate implementation of referenceFS for kerchunked data.

Reply:
Interesting, what is the use-case for it?

Reply:
I believe it is representing data that has this internal structure. Being kerchunk, you are stuck with whatever the original has, rather than being able to rechunk.

cc @d70-t